optuna/pruners/asha.py

```diff
@@ -0,0 +1,145 @@
+from typing import List  # NOQA
+
+
+class ASHAPruner(BasePruner):
+
+    """Pruner using the Asynchronous Successive Halving Algorithm (ASHA).
+
+    ASHA defines rounds (named "rung") started from ``0``.
+    When a trial completed the current rung, it competes with other trials that have completed
+    the same rung. And if it wins the competition, the trial will be promoted to the next rung
+    for continuing the work. Conversely, the losers will be pruned there.
+    This process is repeated by recursively until the trial finishes.
+
+    Please refer to `the original paper <http://arxiv.org/abs/1810.05934>`_
+    for a detailed description of ASHA.
+
+    Note that, unlike the paper, ``ASHAPruner`` recognizes only "number of steps" as the resource
+    consumed by a trial (in the paper, for example, input data size can be treated as a resource).
+    Besides, it does not have a parameter to restrict the maximum resource usage (called ``R`` in
+    the paper). The maximum number of steps executed by a trial is implicitly limited by users via
+    implementation specific parameters (e.g., `step number in simple.py
+    <https://github.com/pfnet/optuna/tree/c5777b3e/examples/pruning/simple.py#L31>`_,
+    `epoch number in chainer_integration.py
+    <https://github.com/pfnet/optuna/tree/c5777b3e/examples/pruning/chainer_integration.py#L65>`_).
+
+    Example:
+
+        We minimize an objective function with ASHA.
+
+        .. code::
+
+            >>> from optuna import create_study
+            >>> from optuna.pruners import ASHAPruner
+            >>>
+            >>> def objective(trial):
+            >>>     ...
+            >>>
+            >>> study = create_study(pruner=ASHAPruner())
+            >>> study.optimize(objective)
+
+    Args:
+        r:
+            A parameter for specifying the minimum resource allocated to a trial.
+            More precisely, a trial is never pruned until it executes ``r * (eta ** s)`` steps
+            (i.e., the completion point of the first rung).
+            When it completes the first rung, it will be promoted to the next rung only if
+            the value of the trial is placed in the top ``1/eta`` fraction of the whole trials that
+            already have reached the point (otherwise it will be pruned there). If the trial won
+            the competition, it continues to execute its work until the next rung
+            completion point (i.e., ``r * (eta ** (s + rung))`` steps) is reached and then repeats
+            the same process with a new ``rung``.
+        eta:
+            A parameter for specifying reduction factor of promotable trials.
+            At the completion point of the each rung, about ``1/eta`` trials will be pruned.
+        s:
+            A parameter for specifying the completion point of the first rung.
+            The point is determined by the expression ``r * (eta ** s)``.
+    """
+
+    def __init__(self, r=1, eta=4, s=0):
```

Comment:
I prefer option (b), as it would make both library and user codes more readable. I appreciate if you could describe the corresponding variable names in the paper in the docstring.